test/controllers/tasks_controller_test.rb

```diff
@@ -1,126 +1,120 @@
-      expect(new_task.due_date.to_time.to_i).must_equal task_hash[:task][:due_date].to_i
-      expect(new_task.completed).must_equal task_hash[:task][:completed]
-
+      expect(new_task.completion_date).must_equal task_hash[:task][:completion_date]
       must_respond_with :redirect
       must_redirect_to task_path(new_task.id)
     end
   end
 
-  # Unskip and complete these tests for Wave 3
-  describe "edit" do
-    it "can get the edit page for an existing task" do
-      skip
-      # Your code here
-    end
-
-    it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
-    end
-  end
-
-  # Uncomment and complete these tests for Wave 3
+  # Tests for Wave 3
   describe "update" do
-    # Note:  If there was a way to fail to save the changes to a task, that would be a great
+        # Note:  If there was a way to fail to save the changes to a task, that would be a great
     #        thing to test.
-    it "can update an existing task" do
-      # Your code here
+    before do
+      @new_task = Task.create(name: "make beds")
+      @task_hash_with_new_data = {
+        task: {
+          name: "edited task",
+          description: "edited task description",
+        },
+      }
+    end
+    it "can update an existing task and redirects to that task's show page" do
+      existing_task = Task.first
+      expect {
+        patch task_path(existing_task.id), params: @task_hash_with_new_data
+      }.wont_change "Task.count"
+      expect(Task.find_by(id: existing_task.id).name).must_equal "edited task"
+      expect(Task.find_by(id: existing_task.id).description).must_equal "edited task description"
+      must_redirect_to task_path
     end
 
-    it "will redirect to the root page if given an invalid id" do
-      # Your code here
+    it "will respond with redirect when attempting to edit a nonexistent task" do
+      patch task_path(5), params: @task_hash_with_new_data
+      must_respond_with :redirect
     end
   end
 
-  # Complete these tests for Wave 4
+  # Tests for Wave 4
   describe "destroy" do
-    # Your tests go here
-
+    it 'decreases Task count by one after deletion' do
+      task = Task.create(name: "take out compost")
+      old = Task.count
+      delete task_path(task.id)
+      expect(Task.count).must_equal (old - 1)
+    end
   end
 
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+    it 'goes from nil to the current date' do
+      task = Task.create(name: "take out compost")
+      put toggle_complete_path(task.id)
+      expect(Task.find_by(id: task.id).completion_date).must_equal Date.today
+      must_redirect_to tasks_path
+    end
+
+    it 'goes from the current date to nil' do
+      task = Task.create(name: "take out compost")
+      put toggle_complete_path(task.id)
+      put toggle_complete_path(task.id)
+      expect(Task.find_by(id: task.id).completion_date).must_equal nil
```

Comment:
`DEPRECATED: Use assert_nil if expecting nil from test/controllers/tasks_controller_test.rb:116. This will fail in Minitest 6.
.DEPRECATED: Use assert_nil if expecting nil from test/controllers/tasks_controller_test.rb:58. This will fail in Minitest 6.`